### PR TITLE
fix constraint bug in beam search, clean up tests

### PR DESCRIPTION
**Pull request by epwalsh**
See https://github.com/allenai/allennlp/pull/5216#discussion_r674154041 for context. @JohnGiorgi @danieldeutsch

### CODE REVIEW

File: tests/nn/beam_search_test.py [hunk 1]
```diff
@@ -788,9 +757,25 @@ def test_repeated_ngram_blocking_end_to_end(self):
...
+        self.beam_search.max_steps = 5
+        expected_top_k = np.array([[1, 2, 3, 4, 5], [1, 2, 4, 3, 5]])
+        expected_log_probs = np.log(
+            np.array([0.4 * 0.3 * 0.3 * 0.2 * 0.1, 0.4 * 0.3 * 0.2 * 0.3 * 0.1])
+        )
+        self._check_results(
+            expected_top_k=expected_top_k,
+            expected_log_probs=expected_log_probs,
+            take_step=step_function,
+        )
```
Comment (epwalsh) [inline tests/nn/beam_search_test.py:773]: This part fails before the fix.

Comment (epwalsh) [reply tests/nn/beam_search_test.py:773]: The rest of the changes to this file are refactoring.

File: allennlp/nn/beam_search.py [hunk 2]
```diff
@@ -1085,7 +1085,9 @@ def _search(
             # dividing by per_node_beam_size gives the ancestor. (Note that this is integer
             # division as the tensor is a LongTensor.)
             # shape: (batch_size, beam_size)
-            backpointer = restricted_beam_indices // self.per_node_beam_size
+            backpointer = torch.divide(
+                restricted_beam_indices, self.per_node_beam_size, rounding_mode="trunc"
+            )
```
Comment (epwalsh) [inline allennlp/nn/beam_search.py:1090]: This avoids a deprecation warning.

File: tests/nn/beam_search_test.py [hunk 3]
```diff
@@ -12,120 +12,98 @@
...
+def get_step_function(
+    transition_matrix: torch.Tensor, with_timestep: bool = False
+) -> Union[StepFunctionTypeNoTimestep, StepFunctionTypeWithTimestep]:
+    def _step_function(
+        last_predictions: torch.Tensor, state: Dict[str, torch.Tensor]
+    ) -> Tuple[torch.Tensor, Dict[str, torch.Tensor]]:
+        log_probs_list = []
+        for last_token in last_predictions:
+            log_probs = torch.log(transition_matrix[last_token.item()])
+            log_probs_list.append(log_probs)
```
Comment (dirkgr) [inline tests/nn/beam_search_test.py:87]: It seems to me that there should be a way to do this without a for loop? Not that it really matters, but something like `Tensor.index_select(...)`?

**Review by dirkgr — APPROVED** [review]

**Pull request merged**